### ISSUE TITLE
generalization ambiguity 2 - even yoke-ier

### ISSUE THREAD

**Issue by lcnr** [rust-lang/trait-system-refactor-initiative#110]
```rust
#![allow(unused)]
trait Yokeable<'a> {
    type Output;
}
trait Id {
    type Refl;
}

fn into_deserialized<M: Id>() -> M
where
    M::Refl: for<'a> Yokeable<'a>,
{
    try_map_project::<M, _>(|_| todo!())
}

fn try_map_project<M: Id, F>(_f: F) -> M
where
    M::Refl: for<'a> Yokeable<'a>,
    F: for<'a> FnOnce(&'a ()) -> <<M as Id>::Refl as Yokeable<'a>>::Output,
{
    todo!()
}

fn main() {}
```

**Comment by lcnr** [comment]
likely caused by closure signature inference

**Comment by lcnr** [comment]
```rust
#![allow(unused)]
trait HigherRanked {
    type Output<'a>;
}
trait Id {
    type Refl: HigherRanked;
}

fn foo<T: Id>() -> for<'a> fn(<<T as Id>::Refl as HigherRanked>::Output<'a>) {
    todo!()
}

fn bar<T: Id>() {
    // generalize here:
    let x = foo::<T>();
}

fn main() {}
```

**Comment by lcnr** [comment]
not caused by closure signature inference: if the outer alias refernces a bound var, the generalizer recurses into it, but then if the nested alias does not refer to bound vars, it gets replaced with an inference variable: 

https://github.com/rust-lang/rust/blob/5ae5d135372c4b576edc73c191d2dc86b1d16b5c/compiler/rustc_infer/src/infer/relate/generalize.rs#L353-L358

We generalize `for<'a> fn(<<T as Id>::Refl as HigherRanked>::Output<'a>)` to `for<'a> fn(<?new as HigherRanked>::Output<'a>)`

**Comment by lcnr** [comment]
fixed by rust-lang#124827
